CDCgov/prime-simplereport
# 5106 Implements validation in device form

# FRONTEND PULL REQUEST

## Related Issue

- Resolves #5106
- Resolves #5184 

## Changes Proposed
- Adds react-hook-forms library to the SimpleReport project. This library will take care of performing basic validations, keep track of the form data state, trigger errors accordingly and perform the focus on error on elements that are properly integrated
- Modifies form common components  `Dropdown, Select, TextInput, MultiSelect` so they can be properly setup with the library.
- Updates the deviceForm component to use `react-hook-form` to track data state, errors and validation. It also removes the prop initialValues as it was not really used.
- Moves the logic of the `disease information` section into its own component.
- Updates necessary unit testing.
- Updates the container components `ManageDeviceTypeFormContainer` and `DeviceTypeFormContainer`
- Adds the role attribute to the component `DeviceTypeReminderMessage`
- Updates the dom references of the multiplex e2e spec 09.

## Additional Information

I ended up modifiying the `MultiSelect` component and the underlying components because the workaround had small glitches.

## Testing
_Tested in Dev1_

- Verify that all fields marked as required must be filled out in order to submit the form.
- Verify that if the field is empty the form should display an error triggered but the click of the `save changes` button.
- Verify that `save changes` button enables when the form is "dirty".
- Verify that `test length` field only allows for numbers and also displays an error if a number outside the 0-999 range is entered.
- Verify that the http request to update the data only triggers if the form has been filled out with valid values.
- Edit mode should load values as expected upon selecting a device. 

## Screenshots / Demos
![Screenshot 2023-03-07 at 4 11 41 PM](https://user-images.githubusercontent.com/103958711/223553853-6d3137cd-0614-41bf-8a5e-3f11dfa7c669.png)
